celery/app/task.py

```diff
@@ -906,11 +906,6 @@ def replace(self, sig):
                 )
 
         if self.request.chain:
-            # We need to freeze the new signature with the current task's ID to
-            # ensure that we don't disassociate the new chain from the existing
-            # task IDs which would break previously constructed results
-            # objects.
-            sig.freeze(self.request.id)
```

Comment:
This may break previous behavior.

We used to have sig.freeze() in #6189 (See https://github.com/celery/celery/pull/6189/files#diff-db604dd7cb51e386710260ff2eba378aac19ba11eec97904bbf097b68caeada6R905) and it didn't fix the problem.

Reply:
I saw the `sig.freeze(self.request.id)` down on L929/932 which is why I removed this one assuming it was a duplicate.

I do see that was also the case when you added the line removed here, back in #6189. My comment at [0] suggested the extra `sig.freeze(self.request.id)` as a fix for the `test_chain_replaced_with_a_chain_and_an_error_callback()` test which was failing prior to it being added. I'll take some time to run through that PR and this one to see if this is truly a duplicate line or if there's something subtle (perhaps the ordering of the `sig.freeze` and the later `sig.set()` on L923?).

[0] https://github.com/celery/celery/pull/6189#pullrequestreview-511455254

Reply:
Per below, I've run through the behaviour of this a bit more and can't really explain why we had the early `sig.freeze()` call here. The test I mentioned in the comment above still runs and succeeds, so either it wasn't really necessary or whatever subtlety previously caused it to be required is now changed/gone. The new reworked diff continues to only `freeze()` the original (to be replaced) task ID into the new signature after everything else is done, and the canvas integration tests appear to function as they do on master (a couple of failures with the retries tests :/).

I plan to come back and take another look over this with fresh eyes to make sure I'm not missing anything, but for now I think it's OK.

Reply:
Scratch that, I've spent some time dumping task IDs again and have come to a more sensible version of the same conclusion I reached last time. `freeze()`ing the signature before we chain any others onto it ensures that the signature we're going to use to replace the original one with will have its ID set correctly, which ensures that the entire chain of parent tasks can be evaluated by the original result object created when the tasks were first applied.

It actually seems to be the second `freeze()` which is redundant, since by that time (when there are tasks to chain onto the replacement signature) `sig` is a chain signature with a final task which has already retained its original task ID. However, that second `freeze()` was necessary because in situation where there were no other tasks to chain onto the replacement signature, the first `freeze()` would not have been called. So the first `freeze()` was critical or it wasn't called, and the second was a no-op or it was critical.

Despite this, the result code (at least for redis) was resilient enough that it was passing the tests anyway!

The new code structure for `replace()` ensures that we only call `freeze()` once and it happens before we chain anything onto the replacement signature. This should be a fine solution. However, I'd like to work out some integration tests which will capture any regressions here since it's difficult to reason about when things seem to work fine despite being subtly broken.

Reply:
The new diff handles this properly now and there are some tests which catch regression if I commentout the single `sig.freeze()` which is done now. I'll leave this or you to resolve if you're happy with it.